### PR TITLE
Preserve pinned flag in {ReadOnly}Memory<T>.Slice

### PR DESCRIPTION
**Pull request by ahsonkhan**
Fixes https://github.com/dotnet/corefx/issues/29239

cc @joshfree, @GrabYourPitchforks, @geoffkizer, @stephentoub, @benaadams  

This change will need to be mirrored to coreclr. cc @Anipik, @safern

### CODE REVIEW

File: src/Common/src/CoreLib/System/Memory.cs [hunk 1]
```diff
@@ -245,13 +248,16 @@ public Memory<T> Slice(int start)
...
+            int capturedLength = _length;
+            int actualLength = capturedLength & RemoveFlagsBitMask;
             if ((uint)start > (uint)actualLength || (uint)length > (uint)(actualLength - start))
             {
                 ThrowHelper.ThrowArgumentOutOfRangeException();
             }
 
-            return new Memory<T>(_object, _index + start, length);
+            // Set the high-bit to match the this._length high bit (1 for pre-pinned, 0 for unpinned).
+            return new Memory<T>(_object, _index + start, length | (capturedLength << 31));
```
Comment (benaadams) [inline src/Common/src/CoreLib/System/Memory.cs]: ` | (capturedLength << 31)` that's low bit?

File: src/Common/src/CoreLib/System/ReadOnlyMemory.cs [hunk 2]
```diff
@@ -167,13 +170,16 @@ public ReadOnlyMemory<T> Slice(int start)
...
+            int capturedLength = _length;
             int actualLength = _length & RemoveFlagsBitMask;
             if ((uint)start > (uint)actualLength || (uint)length > (uint)(actualLength - start))
             {
                 ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.start);
             }
 
-            return new ReadOnlyMemory<T>(_object, _index + start, length);
+            // Set the high-bit to match the this._length high bit (1 for pre-pinned, 0 for unpinned).
+            return new ReadOnlyMemory<T>(_object, _index + start, length | (capturedLength << 31));
```
Comment (benaadams) [inline src/Common/src/CoreLib/System/ReadOnlyMemory.cs]: Same as above

File: src/Common/src/CoreLib/System/Memory.cs (hunk 1 above)
Comment (ahsonkhan) [reply src/Common/src/CoreLib/System/Memory.cs]: Yes, you are correct. Thanks.

**Comment by stephentoub** [comment]
Can some slicing tests be added to https://github.com/dotnet/corefx/blob/master/src/System.Memory/tests/MemoryMarshal/CreateFromPinnedArray.cs?

File: src/Common/src/CoreLib/System/Memory.cs [hunk 3]
```diff
@@ -225,13 +225,16 @@ public override string ToString()
...
+            int capturedLength = _length;
+            int actualLength = capturedLength & RemoveFlagsBitMask;
             if ((uint)start > (uint)actualLength)
             {
                 ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.start);
             }
 
-            return new Memory<T>(_object, _index + start, actualLength - start);
+            // It is expected for (capturedLength - start) to be negative if the memory is already pre-pinned.
+            return new Memory<T>(_object, _index + start, capturedLength - start);
```
[the remark this answers is not in the document]
Comment (GrabYourPitchforks) [reply src/Common/src/CoreLib/System/Memory.cs]: Yup! The reason this works is that `actualLength - start` is always guaranteed non-negative due to the earlier check, so the high bit of `capturedLength` is never treated as a carry bit from an arithmetic perspective. Thus it's preserved for all inputs.

**Comment by ahsonkhan** [comment]
> Can some slicing tests be added to 

I am working on adding such tests, but it requires exposing internals of `Memory<T>` to the test project.

**Pull request merged**